src/dmd/traits.d

```diff
@@ -1694,12 +1694,35 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
         if (dim != 2)
             return dimError(2);
 
-        if (!TemplateInstance.semanticTiargs(e.loc, sc, e.args, 0))
+        // https://issues.dlang.org/show_bug.cgi?id=20761
+        // tiarg semantic may expand the list of arguments leading to such cases:
+        //
+        //     before tiarg sema:  __traits(isSame, seq!(0,0), seq!(1,1))
+        //     after            :  __traits(isSame, 0, 0, 1, 1)
+        //
+        // so we split in two lists and take a new path in case of expansion
+        Objects ob1; ob1.push((*e.args)[0]);
+        Objects ob2; ob2.push((*e.args)[1]);
+
+        if (!TemplateInstance.semanticTiargs(e.loc, sc, &ob1, 0))
+            return new ErrorExp();
+        if (!TemplateInstance.semanticTiargs(e.loc, sc, &ob2, 0))
             return new ErrorExp();
 
+        // example:  __traits(isSame, seq!(0,0), seq(1,1,1)  -> 2 != 3
+        if (ob1.dim != ob2.dim)
+            return False();
+        // expansion has happened
+        else if (ob1.dim > 1)
+        {
+            foreach (immutable i; 0 .. ob1.dim)
+                if (!ob1[i].equals(ob2[i]))
```

Comment:
Don't think this should use `equals` directly and instead apply `isSame` pairwise to all elements:

Example:
```
static assert(__traits(isSame, i => i.value, a => a.value)); // Passes

static assert(__traits(isSame, (a, b) => a + b, (x, y) => x + y)); ; // Passes

static assert(__traits(isSame, // Fails
    Seq!(
        i => i.value,
        (a, b) => a + b
    ),
    Seq!(
        a => a.value,
        (x, y) => x + y
    )
));
```

TL:DR:
`__traits(isSame(Seq!(x0, x1, ...), Seq!(y0, y1, ...)))` should be equal  to
`__traits(isSame(x0, y0)) && __traits(isSame(x1, y1)) && ....`

Reply:
yeah, I know. The current implementation needs to be refactrored... something like bool `isSame (RootObject, RootObject, Scope* sc);`, then I can use this func on each individual element, i.e after expansion.

Reply:
the call to `equals` does not cover the specials cases for each individual element

Reply:
Then i would suggest to split this PR into two commits:
1. Refactor the current logic into a nested function
2. Apply your fix using the introduced function

Reply:
no need to split

Reply:
Refactorings and bug fixes should always be split in different commits s.t. it's easier to track down potential errors later , e.g. when using digger. (Also greatly improves the reviewability of a PR)